ipfs/coreunix/add.go

```diff
@@ -0,0 +1,642 @@
+	balanced "gx/ipfs/QmfB3oNXGGq9S4B2a9YeCajoATms3Zw2VvDm8fK7VeLSV8/go-unixfs/importer/balanced"
+	ihelper "gx/ipfs/QmfB3oNXGGq9S4B2a9YeCajoATms3Zw2VvDm8fK7VeLSV8/go-unixfs/importer/helpers"
+	trickle "gx/ipfs/QmfB3oNXGGq9S4B2a9YeCajoATms3Zw2VvDm8fK7VeLSV8/go-unixfs/importer/trickle"
+)
+
+var log = logging.Logger("coreunix")
+
+// how many bytes of progress to wait before sending a progress update message
+const progressReaderIncrement = 1024 * 256
+
+var liveCacheSize = uint64(256 << 10)
+
+type Link struct {
+	Name, Hash string
+	Size       uint64
+}
+
+type Object struct {
+	Hash  string
+	Links []Link
+	Size  string
+}
+
+// NewAdder Returns a new Adder used for a file add operation.
+func NewAdder(ctx context.Context, p pin.Pinner, bs bstore.GCBlockstore, ds ipld.DAGService) (*Adder, error) {
+	bufferedDS := ipld.NewBufferedDAG(ctx, ds)
+
+	return &Adder{
+		ctx:        ctx,
+		pinning:    p,
+		blockstore: bs,
+		dagService: ds,
+		bufferedDS: bufferedDS,
+		Progress:   false,
+		Hidden:     true,
+		Pin:        true,
+		Trickle:    false,
+		Wrap:       false,
+		Chunker:    "",
+	}, nil
+}
+
+// Adder holds the switches passed to the `add` command.
+type Adder struct {
+	ctx        context.Context
+	pinning    pin.Pinner
+	blockstore bstore.GCBlockstore
+	dagService ipld.DAGService
+	bufferedDS *ipld.BufferedDAG
+	Out        chan<- interface{}
+	Progress   bool
+	Hidden     bool
+	Pin        bool
+	Trickle    bool
+	RawLeaves  bool
+	Silent     bool
+	Wrap       bool
+	Name       string
+	NoCopy     bool
+	Chunker    string
+	root       ipld.Node
+	mroot      *mfs.Root
+	unlocker   bstore.Unlocker
+	tempRoot   cid.Cid
+	CidBuilder cid.Builder
+	liveNodes  uint64
+}
+
+func (adder *Adder) mfsRoot() (*mfs.Root, error) {
+	if adder.mroot != nil {
+		return adder.mroot, nil
+	}
+	rnode := unixfs.EmptyDirNode()
+	rnode.SetCidBuilder(adder.CidBuilder)
+	mr, err := mfs.NewRoot(adder.ctx, adder.dagService, rnode, nil)
+	if err != nil {
+		return nil, err
+	}
+	adder.mroot = mr
+	return adder.mroot, nil
+}
+
+// SetMfsRoot sets `r` as the root for Adder.
+func (adder *Adder) SetMfsRoot(r *mfs.Root) {
+	adder.mroot = r
+}
+
+// Constructs a node from reader's data, and adds it. Doesn't pin.
+func (adder *Adder) add(reader io.Reader) (ipld.Node, error) {
+	chnk, err := chunker.FromString(reader, adder.Chunker)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure all added nodes are written when done.
+	defer adder.bufferedDS.Commit()
+
+	params := ihelper.DagBuilderParams{
+		Dagserv:    adder.bufferedDS,
+		RawLeaves:  adder.RawLeaves,
+		Maxlinks:   ihelper.DefaultLinksPerBlock,
+		NoCopy:     adder.NoCopy,
+		CidBuilder: adder.CidBuilder,
+	}
+
+	if adder.Trickle {
+		return trickle.Layout(params.New(chnk))
+	}
+
+	return balanced.Layout(params.New(chnk))
+}
+
+// RootNode returns the root node of the Added.
+func (adder *Adder) RootNode() (ipld.Node, error) {
+	// for memoizing
```

Comment:
I personally like the word "memoize", but instead this should explain what it means. Such as "Root node never changes, so only calculate it once, then return that same result on future calls".